zilliqa/src/p2p_node.rs

```diff
@@ -255,12 +251,37 @@ impl P2pNode {
                         SwarmEvent::NewListenAddr { address, .. } => {
                             info!(%address, "P2P swarm listening on");
                         }
+                        // this is necessary - https://docs.rs/libp2p-kad/latest/libp2p_kad/#important-discrepancies
+                        SwarmEvent::Behaviour(BehaviourEvent::Identify(identify::Event::Received { peer_id, info, .. })) => {
+                            debug!(%peer_id, ?info, "identify event");
+                            if info.protocols.iter().any(|p| *p == kad::PROTOCOL_NAME) {
+                                for addr in info.listen_addrs {
+                                    // this will trigger the `NewExternalAddrOfPeer` event below
+                                    self.swarm.add_peer_address(peer_id, addr);
+                                }
+                            }
+                        }
+                        SwarmEvent::ExternalAddrExpired{address} => {
+                            debug!(%address, "expired");
+                            self.swarm.remove_external_address(&address);
+                        }
+                        SwarmEvent::NewExternalAddrOfPeer{peer_id, address} => {
+                            debug!(%peer_id, %address, "new peer");
+                            self.swarm.behaviour_mut().kademlia.add_address(&peer_id, address);
+                            self.swarm.behaviour_mut().gossipsub.add_explicit_peer(&peer_id);
```

Comment:
This adds the address as an 'explicit peer' in Gossipsub, which means it ignores lots of the gossipsub logic around scoring, etc.

Is that intended?

`gossipsub::Behaviour` already adds 'normal' peers when a connection is established:
https://github.com/libp2p/rust-libp2p/blob/88f7875ad1a3e240aa2d9b9fb6f6c5354f1a62eb/protocols/gossipsub/src/behaviour.rs#L2888